pkg/cloudlogging/puller_test.go

```diff
@@ -115,23 +221,62 @@ func setupFakeClient(t *testing.T, ctx context.Context, s *fakeServer) *logging.
 	t.Cleanup(func() {
 		conn.Close()
 	})
-	// fakeClient := loggingpb.NewLoggingServiceV2Client(conn)
+
 	fakeClient, err := logging.NewClient(ctx, option.WithGRPCConn(conn))
 	if err != nil {
 		t.Fatalf("creating client for fake at %q: %v", addr, err)
 	}
 	return fakeClient
 }
 
+func setupFakeStreamClient(t *testing.T, ctx context.Context, s *fakeServer) (*logging.Client, loggingpb.LoggingServiceV2_TailLogEntriesClient) {
+	t.Helper()
+
+	// Setup fake server.
+	addr, conn := testutil.FakeGRPCServer(t, func(grpcS *grpc.Server) {
+		loggingpb.RegisterLoggingServiceV2Server(grpcS, s)
+	})
+	t.Cleanup(func() {
+		conn.Close()
+	})
+
+	fakeClient, err := logging.NewClient(ctx, option.WithGRPCConn(conn))
+	if err != nil {
+		t.Fatalf("creating client for fake at %q: %v", addr, err)
+	}
+	tailLogEntriesClient, _ := fakeClient.TailLogEntries(ctx)
+	return fakeClient, tailLogEntriesClient
+}
+
 type fakeServer struct {
 	loggingpb.UnimplementedLoggingServiceV2Server
 
-	req         *loggingpb.ListLogEntriesRequest
-	resp        *loggingpb.ListLogEntriesResponse
+	listReq     *loggingpb.ListLogEntriesRequest
+	listResp    *loggingpb.ListLogEntriesResponse
+	tailReq     *loggingpb.TailLogEntriesRequest
+	tailResp    *loggingpb.TailLogEntriesResponse
```

Comment:
It might make sense to have a counter and each tail call we just send one log back and increase the counter until a cap

Reply:
Added

Reply:
That's not what I meant. I meant instead of providing the logs via `tailResp *loggingpb.TailLogEntriesResponse`, you can generate log on the fly. See my other comment.